src/LazyJwtCredentialsManager.ts

```diff
@@ -0,0 +1,65 @@
+import { CredentialsManager } from './CredentialsManager';
+import { JwtCredentials } from './JwtCredentialsManager';
+
+const SECONDS_TO_MS = 1000;
+
+export default abstract class LazyJwtCredentialsManager<
+  TCredentials extends JwtCredentials
+> implements CredentialsManager<TCredentials> {
+  private token: string | null | undefined;
+
+  private credentialsPromise: Promise<TCredentials | null | undefined> | null;
+
+  constructor() {
+    this.token = null;
+    this.credentialsPromise = null;
+  }
+
+  private isExpired(credentials: TCredentials) {
+    return Date.now() >= credentials.exp * SECONDS_TO_MS;
+  }
+
+  async getCredentials(): Promise<TCredentials | null | undefined> {
+    const credentials = await this.credentialsPromise;
+
+    if (credentials && this.isExpired(credentials)) {
```

Comment:
uhm, so this acts as a lock, basically? trying to understand if there are instances where `updateCredentials` could be called twice

Reply:
the stored promise is some manner of lock, it's not great one but I think it's fine if updateCredentials is called multiple times?

Reply:
I think so!